### ISSUE TITLE
Is appSession persisted?

### ISSUE THREAD

**Issue by JaneJeon** [auth0/express-openid-connect#210]
Hi again, I found another hole in the documentation that might be make or break for me.

In the documentation, `afterCallback` is noted as "Function for custom callback handling after receiving tokens and before redirecting", and the example there shows persisting `userProfile` to the `appSession`.

However, does that `userProfile` (attached to the `appSession`) only persist for that specific request and gets lost from there on? Or can I access `req.appSession.userProfile` on every request from then on from that user?

The reason I'm asking about this is because I want a way to "replace" `req.oidc.user` with something custom. In particular, I am using `afterCallback` to create an instance of User model (and possibly provision them in my own DB if needed) and persist it so that in my session, I can refer to the instance of user on my service's database, NOT the one provided by OIDC (think about it as typical SSO authentication -> provisioning user locally).

Basically, I'm trying to use this as a passport strategy (because the underlying library, `openid-client`'s passport strategy documentation is nil) and how `appSession` is "hydrated" on each request would drastically change my implementation.

If what I'm trying to do is impossible with this library, or if there's some more obvious, easier way of doing local provisioning that I'm missing, please let me know!

(I'm REALLY sorry for making so many issues about documentation-related stuff, but the documentation around `req.oidc.user`, `req.appSession`, and basic workflow such as SSO auth -> local provisioning is very thin :/)

**Comment by adamjmcgrath** [comment]
Hi @JaneJeon - no need to apologise, always happy to get feedback :)

> However, does that userProfile (attached to the appSession) only persist for that specific request and gets lost from there on? Or can I access req.appSession.userProfile on every request from then on from that user?

Anything you add to the session will be persisted on future requests. 

> The reason I'm asking about this is because I want a way to "replace" req.oidc.user with something custom.

You can't replace `req.oidc.user` as this is a getter that decodes the user's claims from the ID Token, see https://github.com/auth0/express-openid-connect/blob/master/lib/context.js#L125-L147

> Basically, I'm trying to use this as a passport strategy

This SDK is designed as an alternative to Passport.js, not as a Passport strategy - so you may struggle to use it as a Passport strategy. Personally I would either use this as an alternative or persist with https://github.com/panva/node-openid-client/blob/main/docs/README.md#strategy

**Comment by JaneJeon** [comment]
Oh, that's unfortunate. (What I meant by "passport replacement" isn't that I LITERALLY wanted a passport strategy, but some way to persist the user locally).

With all of this, would you kindly make a note in the documentation that this library is more suited for using OIDC-authenticated users DIRECTLY, rather than authenticating users using OIDC and provisioning them locally?

I think most people, when they work with OIDC, think of "provisioning" users in the latter way - for example, that's the way GitHub, Slack, and basically most of the SSO-enabled apps work.

So it would really save others from going down the wrong rabbit hole as I did and wasting a whole bunch of time trying to fit this library to do what I thought was a "supported" use case.

Thanks

**Comment by adamjmcgrath** [comment]
Hi @JaneJeon 

You can still provision a user locally using the `afterCallback` example https://auth0.github.io/express-openid-connect/interfaces/configparams.html#aftercallback

In that example `req.appSession.userProfile` will be persisted for the duration of the session
